Add functions and tests for parity between Turtle System (Python) and Turtle Module

LordSaumya opened the pull request.

LordSaumya committed:
Create scaling method for simplicity

LordSaumya committed:
Add inverse trigonometric functions

LordSaumya committed:
Refactor to use helper functions

LordSaumya committed:
Add checkfile and checkdir functions

LordSaumya committed:
Add trigonometric functions

LordSaumya committed:
Fix delete command

LordSaumya committed:
Add checkdir command

LordSaumya committed:
Add console command

LordSaumya committed:
Add eof and eoln functions

LordSaumya committed:
Add remaining maths functions (exp, pow, log, sqrt, pi)

LordSaumya committed:
Add file handling commands

LordSaumya committed:
Add the find commands

LordSaumya committed:
Add the hypot command

LordSaumya committed:
Add isfile and isdir commands

LordSaumya committed:
Add sign command

LordSaumya committed:
Add keybuffer and keyecho commands

LordSaumya committed:
Add qval function

LordSaumya committed:
Add read function for keybuffer

LordSaumya committed:
Add functions to make and remove dirs and files

LordSaumya committed:
Add functions for random integers

LordSaumya committed:
Add root function

LordSaumya committed:
Add ord function

LordSaumya committed:
Add recolour function

LordSaumya committed:
Add time functions

LordSaumya committed:
Prefix helpers with underscore

LordSaumya committed:
Add appropriate type aliases for files and handles

LordSaumya committed:
Add tests for helper functions

LordSaumya committed:
Fix error in divmult

LordSaumya committed:
Remove ord since it is already present in python

LordSaumya committed:
Add tests for maths functions

LordSaumya committed:
Add tests for canvas setup functions

LordSaumya committed:
Add tests for state management functions

LordSaumya committed:
Add tests for string and conversion functions

LordSaumya committed:
Fix qval assertions

LordSaumya committed:
Add tests for console, system, and time functions

LordSaumya committed:
Add tests for random and colour functions

LordSaumya committed:
Fix time function

LordSaumya committed:
Add tests for drawing configuration functions

LordSaumya committed:
Fix tests for randomness

LordSaumya committed:
Add tests for movement

LordSaumya committed:
Change drawing functions to return an ID for testing and consistency

LordSaumya committed:
Add tests for drawing functions

LordSaumya committed:
Add tests for event and input handling functions

LordSaumya committed:
Fix file handling functions

LordSaumya committed:
Add tests for file and directory handling functions

ioanav104 commented on tests/file_handling/test_file_dir_handling.py:
```diff
@@ -0,0 +1,939 @@
+TclError = tkinter.TclError  # in case needed
+
+# Use a lambda to check if two floats are close (avoiding floating point errors)
+is_close = lambda a, b: abs(a - b) < 1e-6
```
You could also use `math.isclose`: https://docs.python.org/3/library/math.html#math.isclose

ioanav104 commented on tests/file_handling/test_file_dir_handling.py:
```diff
@@ -0,0 +1,939 @@
+
+# Fixture to create a temporary file for testing
+@pytest.fixture
+def temp_file(tmp_path):
```
There's already a standard fixture for this in pytest: https://docs.pytest.org/en/stable/how-to/tmp_path.html

ioanav104 commented on turtle_oxford.py:
```diff
@@ -871,11 +918,750 @@ def qint(s: str, mult: int, default: int) -> int:
+        width=0,
+    )
+
+def randint(a: int, b: int) -> int:
```
Is this necessary?

ioanav104 commented on turtle_oxford.py:
```diff
@@ -871,11 +918,750 @@ def qint(s: str, mult: int, default: int) -> int:
+    """
+    return random.randint(a, b)
+
+def randrange(range: int) -> int:
```
ditto

ioanav104 commented on turtle_oxford.py:
```diff
@@ -871,11 +918,750 @@ def qint(s: str, mult: int, default: int) -> int:
+    """
+    return random.randrange(range)
+
+def randseed(seed: int):
```
Users can just import the `random` library

ioanav104 commented on turtle_oxford.py:
```diff
@@ -871,11 +918,750 @@ def qint(s: str, mult: int, default: int) -> int:
+    try:
+        open(name, "w").close()
+        return True
+    except Exception:
```
open will return `True` for existing files, too, if the user running the process has write permissions to it. Also, this effectively silences the error, which is bad practice because it's impossible to debug. Try redirecting the exception message to stdout or stderr.

ioanav104 commented on turtle_oxford.py:
```diff
@@ -871,11 +918,750 @@ def qint(s: str, mult: int, default: int) -> int:
+    try:
+        os.rmdir(name)
+        return True
+    except Exception:
```
same here with the silent exception

ioanav104 commented on turtle_oxford.py:
```diff
@@ -871,11 +918,750 @@ def qint(s: str, mult: int, default: int) -> int:
+    try:
+        os.mkdir(name)
+        return True
+    except Exception:
```
same here with the silent exception

ioanav104 commented on turtle_oxford.py:
```diff
@@ -871,11 +918,750 @@ def qint(s: str, mult: int, default: int) -> int:
+    :return: the string read from the keyboard buffer
+    :rtype: str
+    """
+    if len(TurtleCanvas._key_buffer) == 0:
```
```suggestion
    if not TurtleCanvas._key_buffer:
```

ioanav104 commented on turtle_oxford.py:
```diff
@@ -871,11 +918,750 @@ def qint(s: str, mult: int, default: int) -> int:
+
+"""A type alias for a mutable handle to store the index of the found directory or file.
+"""
+FindHandle: TypeAlias = list[int]
```
I understand that FindHandle is a workaround to making an int into a mutable type. It took me a while to figure that out. Can you include in the comment that this is only meant to be a singleton and a workaround to not having a mutable int?

ioanav104 commented on turtle_oxford.py:
```diff
@@ -871,11 +918,750 @@ def qint(s: str, mult: int, default: int) -> int:
+    :rtype: list[str]
+    """
+    pattern = os.path.join(os.getcwd(), pattern)
+    return glob.glob(pattern)
```
```suggestion
    return glob.glob(pattern, root_dir=os.getcwd)
```

ioanav104 commented on turtle_oxford.py:
```diff
@@ -871,11 +918,750 @@ def qint(s: str, mult: int, default: int) -> int:
+    """
+    file_handle.write(data)
+
+def fwriteline(file_handle: file, data: str):
```
Can you name this variable something else? It's confusing with the FileHandle type and the file_handle variables of the FileHandle type.

ioanav104 commented on turtle_oxford.py:
```diff
@@ -871,11 +918,750 @@ def qint(s: str, mult: int, default: int) -> int:
+        shutil.copy(file_name, copy_name)
+        return open(copy_name, mode)
+    except Exception as e:
+        raise RuntimeError("Copy failed") from e
```
Try not to obscure the original exception message

ioanav104 commented on turtle_oxford.py:
```diff
@@ -871,11 +918,750 @@ def qint(s: str, mult: int, default: int) -> int:
+    file_handle.seek(curr)
+    return ch == "\n"
+
+def exp(x: float) -> float:
```
Are all of these calls to the math functions really needed?

ioanav104 commented on turtle_oxford.py:
```diff
@@ -871,11 +918,750 @@ def qint(s: str, mult: int, default: int) -> int:
+    new_code = code | (bit6 << 6) | (bit7 << 7)
+    return format(new_code, '08b')
+
+def checkdir(pcode: str, dir_name: str) -> str:
```
This function is quite lengthy, it could benefit from a docstring

ioanav104 commented on tests/test_maths.py:
```diff
@@ -0,0 +1,121 @@
+    # Test antilog: 10^(8/4) * 5 = 500
+    assert is_close(antilog(8, 4, 5), 500)
+
+def test_cos(tc_setup):
```
I wouldn't add tests for the functions coming straight from `math`, you're testing a third party library.

ioanav104 commented on tests/test_events_inputs.py:
```diff
@@ -0,0 +1,208 @@
+is_close = lambda a, b: abs(a - b) < 1e-6
+
+# Fixture to create a canvas and override mainloop so tests don't block
+@pytest.fixture
```
This seems common to a lot of tests, you could define it in a test_utils.py module which you can import for testing purposes.

ioanav104 commented on tests/test_events_inputs.py:
```diff
@@ -0,0 +1,208 @@
+
+# The detect test has been commented out because it has indeterministic behaviour due to threading
+
+# def test_detect(monkeypatch, tc_setup):
```
I may have some time to play with this next week, but until then, you may want to have a look at multiprocessing instead of threading: https://docs.python.org/3/library/multiprocessing.html

It's more pythonic and you can start the processes at the same time and join them at the same time.
You may also want to have a look at pytest-repeat: https://pypi.org/project/pytest-repeat/
It should help you run this test 20 times in a row to see if it's still flaky.

ioanav104 commented on tests/test_colours_random.py:
```diff
@@ -0,0 +1,50 @@
+    # Test mixcols: (10 * 2 + 20 * 5)//7 = 16
+    assert mixcols(10, 20, 2, 5) == 17
+
+def test_randint(tc_setup):
```
The next 3 seem to just be testing a third party library, I don't think they're worth including.

ioanav104 commented on tests/test_canvas_setup.py:
```diff
@@ -0,0 +1,36 @@
+        yield tc
+
+def test_update(tc_setup):
+    update()
```
These are testing the internal state, not behaviour, which is generally bad practice. You're not really catching any potential mistakes with these tests, only making future refactors harder.

ioanav104 commented on tests/file_handling/test_file_dir_handling.py:
```diff
@@ -0,0 +1,939 @@
+# Tests for input validation
+def test_checkfile_pcode_length():
+    # Test with invalid PCode length
+    with pytest.raises(ValueError, match="PCode must be 8 characters long"):
```
This is debatable, but I would argue you don't need to match the exact error message here as a simple tweak to the error message shouldn't require an edit to the tests if the behaviour is largely unchanged.

ioanav104 commented on tests/file_handling/test_file_dir_handling.py:
```diff
@@ -0,0 +1,939 @@
+        checkfile("0123abcd", "test.txt")
+
+# Tests for action 0: Inquiry only
+def test_checkfile_action0_file_exists(monkeypatch):
```
These tests are very similar, so you could keep the code DRYer by parameterising the tests: https://docs.pytest.org/en/stable/how-to/parametrize.html

ioanav104 commented on tests/file_handling/test_file_dir_handling.py:
```diff
@@ -0,0 +1,939 @@
+    monkeypatch.setattr(os, "chdir", lambda _: None)
+    
+    # Test chdir
+    chdir("test_dir")
```
What behaviour is being tested here and what potential mistakes are caught?

ioanav104 reviewed (commented):
Thanks for putting in all this effort, it looks like hard work. Sorry it took me a week to take a look.
Overall the code looks very good!
I did have some comments, some nit-picks. In particular I don't understand why a lot of functions third party libraries from `random` and `math` are included and they can't just be used directly from those libraries. I do understand the reimplementation of the file and folder functions, as Peter does those quite differently in his version of Turtle Python. I can be convinced otherwise though, don't be afraid to push back where appropriate. I can have another look in a few days.

LordSaumya commented:
Hi @ioanav104, thank you for the review, I will look through the suggested changes and implement them soon. 

Regarding the third-party library functions (such as `random` and `math`), Prof. Millican discussed in our meeting that I should add them to the library for parity with Turtle System.

pjrmillican commented:
Thank you both for your work on this!  Saumya, would you be free to chat some time soon compatibly with Singapore timings, either this week or at the weekend?
All the best,
Peter

From: Saumya Shah ***@***.***>
Sent: 17 March 2025 19:17
To: turtleoxford/turtle-python ***@***.***>
Cc: pjrmillican ***@***.***>; Review requested ***@***.***>
Subject: Re: [turtleoxford/turtle-python] Add functions and tests for parity between Turtle System (Python) and Turtle Module (PR #13)


Hi @ioanav104<https://github.com/ioanav104>, thank you for the review, I will look through the suggested changes and implement them soon.

Regarding the third-party library functions (such as random and math), Prof. Millican discussed in our meeting that I should add them to the library for parity with Turtle System.

—
Reply to this email directly, view it on GitHub<https://github.com/turtleoxford/turtle-python/pull/13#issuecomment-2729114377>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALNVW5SPGHQF2A257HAJH4D2U2VJTAVCNFSM6AAAAABYSALSEGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRZGEYTIMZXG4>.
You are receiving this because your review was requested.Message ID: ***@***.******@***.***>>
[LordSaumya]LordSaumya left a comment (turtleoxford/turtle-python#13)<https://github.com/turtleoxford/turtle-python/pull/13#issuecomment-2729114377>

Hi @ioanav104<https://github.com/ioanav104>, thank you for the review, I will look through the suggested changes and implement them soon.

Regarding the third-party library functions (such as random and math), Prof. Millican discussed in our meeting that I should add them to the library for parity with Turtle System.

—
Reply to this email directly, view it on GitHub<https://github.com/turtleoxford/turtle-python/pull/13#issuecomment-2729114377>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALNVW5SPGHQF2A257HAJH4D2U2VJTAVCNFSM6AAAAABYSALSEGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRZGEYTIMZXG4>.
You are receiving this because your review was requested.Message ID: ***@***.******@***.***>>

LordSaumya committed:
Add docstrings to all functions

LordSaumya committed:
Fix update tests to test for refresh calls and number of shapes on canvas

LordSaumya committed:
Parametrise checkdir/checkfile tests

LordSaumya committed:
Fix chdir test

LordSaumya committed:
Add test utilities

LordSaumya committed:
Change all test files to use test utilities

LordSaumya committed:
Fix comments for variable names and silent exceptions

LordSaumya committed:
Add alternative version of Turtle Oxford with maths functions closer to the original Turtle System

LordSaumya commented:
Hi,

@ioanav104 Thanks for the comments, I have gone through them and fixed some of the issues in the tests and the library. I looked into multiprocessing for the `detect` test, but I don't think it works because testing `detect` requires a shared memory space between the process running `detect` and the process simulating the keystroke. After discussion with Prof Millican, I have kept the third-party library functions for parity.

@pjrmillican As discussed, I have added the `strict` version of the library with maths functions syntactically closer to the original TurtleSystem. I have also added docstrings to all the functions which makes it easy to generate a full documentation using Sphinx.

LordSaumya committed:
Fix dir and file finder helper function

LordSaumya committed:
Add plus version of turtle with layers (z coordinates)

LordSaumya committed:
Fix description

LordSaumya committed:
Add Turtle class and layer methods

LordSaumya committed:
Add documentation files

LordSaumya committed:
Add examples

LordSaumya committed:
Fix tests

LordSaumya committed:
Update toml, lock, and gitignore

LordSaumya committed:
Fix documentation and commands

LordSaumya committed:
Add scripts to run examples

LordSaumya committed:
Update example documentation
